test/contract.errors.js

```diff
@@ -0,0 +1,166 @@
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+    })
+
+    it('errors when a non-existant event is listened for', function(){
+        const expected = 'Event "void" doesn\'t exist in this contract';
+        const contract = new eth.Contract(abi);
+
+        try {
+            contract.once('void', () => {});
+            assert.fail();
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+    })
+
+    it('errors when listening for event without setting address', function(){
+        const expected = 'This contract object doesn\'t have address set yet, ' +
+                         'please set an address first';
+
+        const eventAbi = [{
+            "anonymous": false,
+            "inputs": [],
+            "name": "BasicEvent",
+            "type": "event"
+        }];
+
+        const contract = new eth.Contract(eventAbi);
+
+        try {
+            contract.once('BasicEvent', () => {});
+            assert.fail();
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+    })
+
+    it('errors when an event name is reserved', function(){
+        const expected = 'The event "newListener" is a reserved event name, ' +
+                         'you can\'t use it.';
+
+        const newListenerEventAbi = [{
+            "anonymous": false,
+            "inputs": [],
+            "name": "newListener",
+            "type": "event"
+        }];
+
+        const contractAddress = '0xEE221E529cF6DB20179E7bAeb4442e9CbdCa83d7';
+        const contract = new eth.Contract(newListenerEventAbi, contractAddress);
+
+        try {
+            contract.once('newListener', () => {});
+            assert.fail();
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+    })
+
+    it('errors when deploy is called without data', function(done){
+        const expected = 'No "data" specified in neither the given options, ' +
+                         'nor the default options.';
+
+        const newAbi = [{
+          "inputs": [],
+          "payable": false,
+          "stateMutability": "nonpayable",
+          "type": "constructor"
+        }];
+
+        const contract = new eth.Contract(newAbi);
+
+        // Thrown error format
+        try {
+            contract.deploy();
+            assert.fail();
+        } catch(err){
+            assert(err.message.includes(expected));
+        }
+
+        // Callback format
+        contract.deploy({}, function(err){
+            assert(err.message.includes(expected));
+            done();
+        });
+    })
+
+    it('errors when send is called without a *contract address* set', function(){
+        const expected = 'This contract object doesn\'t have address set yet, ' +
+                         'please set an address first.';
+
+        const contract = new eth.Contract(abi);
+        try {
+            contract.methods.simpleMethod("0xaaa").send();
```

Comment:
Can you add the callback and promise case as separate test cases?

Reply:
This one is directly thrown, rather than passed to utils._fireError

Reply:
You would also create two test cases if we would test the ``utils._fireError`` function. One test case one ``it``.

Reply:
This problem is detected before anything gets routed to a callback or promise.